Fix value range and lost params for GPU tonemap filters

gnattu opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://docs.jellyfin.org/general/contributing/issues.html page.
-->

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->

For BT.2390 EETF, the mapped output might get overflowed due to the input signal has a too high value. For this tone mapping operator, the input signal exceeds the input peak should be clipped to the specified input peak value to prevent artifacts due to overflowing.

For ICtCp tone mapper, this means the input RGB color channels has to be clipped to the input peak before color conversion.

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

Fixes #552

gnattu committed:
avfilter/tonemap_*: clip input range for bt2390

gnattu commented:
Pipeline failure is unrelated and is broken before this PR:


```
2025-03-21T06:27:16.1914196Z #31 1.112 -- Check for working CXX compiler: /opt/ct-ng/bin/aarch64-ffbuild-linux-gnu-g++ - broken
2025-03-21T06:27:16.1915367Z #31 1.113 CMake Error at /usr/share/cmake-3.28/Modules/CMakeTestCXXCompiler.cmake:60 (message):
2025-03-21T06:27:16.1916082Z #31 1.113   The C++ compiler
2025-03-21T06:27:16.1916573Z #31 1.113 
2025-03-21T06:27:16.1916978Z #31 1.113     "/opt/ct-ng/bin/aarch64-ffbuild-linux-gnu-g++"
2025-03-21T06:27:16.1917725Z #31 1.113 
2025-03-21T06:27:16.1919892Z #31 1.113   is not able to compile a simple test program.
2025-03-21T06:27:16.1931735Z #31 1.113 
2025-03-21T06:27:16.1932095Z #31 1.113   It fails with the following output:
2025-03-21T06:27:16.1932679Z #31 1.113 
2025-03-21T06:27:16.1933223Z #31 1.113     Change Dir: '/stage/Ne10/build/CMakeFiles/CMakeScratch/TryCompile-GvnHNF'
2025-03-21T06:27:16.1933869Z #31 1.113     
2025-03-21T06:27:16.1934491Z #31 1.113     Run Build Command(s): /usr/bin/cmake -E env VERBOSE=1 /usr/bin/gmake -f Makefile cmTC_99dde/fast
2025-03-21T06:27:16.1935977Z #31 1.113     /usr/bin/gmake  -f CMakeFiles/cmTC_99dde.dir/build.make CMakeFiles/cmTC_99dde.dir/build
2025-03-21T06:27:16.1937077Z #31 1.113     gmake[1]: Entering directory '/stage/Ne10/build/CMakeFiles/CMakeScratch/TryCompile-GvnHNF'
2025-03-21T06:27:16.1937970Z #31 1.113     Building CXX object CMakeFiles/cmTC_99dde.dir/testCXXCompiler.cxx.o
2025-03-21T06:27:16.1940234Z #31 1.113     /opt/ct-ng/bin/aarch64-ffbuild-linux-gnu-g++ --sysroot=/opt/ct-ng/aarch64-ffbuild-linux-gnu/sysroot   -static-libgcc -static-libstdc++ -I/opt/ffbuild/include -O2 -pipe -march=armv8-a -fPIC -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pthread -fvisibility=hidden -fno-semantic-interposition  -o CMakeFiles/cmTC_99dde.dir/testCXXCompiler.cxx.o -c /stage/Ne10/build/CMakeFiles/CMakeScratch/TryCompile-GvnHNF/testCXXCompiler.cxx
2025-03-21T06:27:16.1943538Z #31 1.113     Linking CXX executable cmTC_99dde
2025-03-21T06:27:16.1944245Z #31 1.113     /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_99dde.dir/link.txt --verbose=1
2025-03-21T06:27:16.1948194Z #31 1.113     /opt/ct-ng/bin/aarch64-ffbuild-linux-gnu-g++ --sysroot=/opt/ct-ng/aarch64-ffbuild-linux-gnu/sysroot -static-libgcc -static-libstdc++ -I/opt/ffbuild/include -O2 -pipe -march=armv8-a -fPIC -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pthread -fvisibility=hidden -fno-semantic-interposition  -static-libgcc -static-libstdc++ -L/opt/ffbuild/lib -O2 -pipe -march=armv8-a -fstack-protector-strong -Wl,-z,relro,-z,now -pthread -lm  -rdynamic CMakeFiles/cmTC_99dde.dir/testCXXCompiler.cxx.o -o cmTC_99dde 
2025-03-21T06:27:16.1952627Z #31 1.113     /opt/ct-ng/lib/gcc/aarch64-ffbuild-linux-gnu/14.2.0/../../../../aarch64-ffbuild-linux-gnu/bin/ld: cannot find -lstdc++: No such file or directory
2025-03-21T06:27:16.1953716Z #31 1.113     collect2: error: ld returned 1 exit status
2025-03-21T06:27:16.1954373Z #31 1.113     gmake[1]: *** [CMakeFiles/cmTC_99dde.dir/build.make:99: cmTC_99dde] Error 1
2025-03-21T06:27:16.1955304Z #31 1.113     gmake[1]: Leaving directory '/stage/Ne10/build/CMakeFiles/CMakeScratch/TryCompile-GvnHNF'
2025-03-21T06:27:16.1956120Z #31 1.113     gmake: *** [Makefile:127: cmTC_99dde/fast] Error 2
```

It looks like the updated crosstool-ng is breaking stdc++ linking

nyanmisaka commented on debian/patches/0004-add-cuda-tonemap-impl.patch:
```diff
@@ -1324,8 +1325,11 @@ Index: FFmpeg/libavfilter/cuda/tonemap.cu
-+    float3 ictcp = lrgb2ictcp(rgb.x, rgb.y, rgb.z);
 +    float peak = src.peak;
++    float r = mix(rgb.x, min(rgb.x, peak), tonemap_func == TONEMAP_BT2390);
++    float g = mix(rgb.x, min(rgb.y, peak), tonemap_func == TONEMAP_BT2390);
++    float b = mix(rgb.x, min(rgb.z, peak), tonemap_func == TONEMAP_BT2390);
```
```suggestion
+    float g = mix(rgb.y, min(rgb.y, peak), tonemap_func == TONEMAP_BT2390);
+    float b = mix(rgb.z, min(rgb.z, peak), tonemap_func == TONEMAP_BT2390);
```

typo?

gnattu committed:
avfilter/tonemap_cuda: fix typo

gnattu committed:
avfilter/tonemap_*: fix lost tonemap param during dovi re-init

For certain video inputs, the dolby vision metadata could be
missing on the first few frames which would cause the filter to
re-initialize itself for the first frame containing the dovi metadata.

This re-init behavior would mistakenly re-interpret the current
set param as user input which will create unwanted results due to
wrong tone mapping parameters. Fix this by using a separate value
for calculated param and never change the user input.

gnattu commented:
Updated to fix the lost tonemap params issue for certain video inputs, see https://github.com/jellyfin/jellyfin-ffmpeg/issues/552#issuecomment-2743553057

For certain video inputs, the dolby vision metadata could be missing on the first few frames which would cause the filter to re-initialize itself for the first frame containing the dovi metadata.

This re-init behavior would mistakenly re-interpret the current set param as user input and mistakenly re-calculate the parameter, which will create unwanted results due to wrong tone mapping parameters. Fix this by using a separate value for calculated param and never change the user input. This affects all GPU based tonemap filters.

The CPU based tonemapx filter is not affected because the value is never re-calculated. Instead, it only re-map the dovi metadata.

nyanmisaka commented on debian/patches/0004-add-cuda-tonemap-impl.patch:
```diff
@@ -1935,6 +1936,7 @@ Index: FFmpeg/libavfilter/vf_tonemap_cuda.c
 +static av_cold int init(AVFilterContext *ctx)
 +{
 +    TonemapCUDAContext *s = ctx->priv;
++    s->final_param = NAN;
```
Move to `preinit`?

nyanmisaka commented on debian/patches/0004-add-cuda-tonemap-impl.patch:
```diff
@@ -2347,26 +2349,26 @@ Index: FFmpeg/libavfilter/vf_tonemap_cuda.c
-+    if (isnan(s->param))
-+        s->param = 1.0f;
++    if (isnan(s->final_param))
++        s->final_param = 1.0f;
```
It's not passed to the kernel.

nyanmisaka commented on debian/patches/0007-add-bt2390-eetf-and-code-refactor-to-opencl-tonemap.patch:
```diff
@@ -2460,3 +2481,12 @@ Index: FFmpeg/libavfilter/vf_tonemap_opencl.c
+     .description    = NULL_IF_CONFIG_SMALL("Perform HDR to SDR conversion with tonemapping."),
+     .priv_size      = sizeof(TonemapOpenCLContext),
+     .priv_class     = &tonemap_opencl_class,
+-    .init           = &ff_opencl_filter_init,
```
Use `AVFilter.preinit =` instead?

nyanmisaka commented on debian/patches/0052-add-vf-tonemap-videotoolbox-filter.patch:
```diff
@@ -1238,7 +1239,8 @@ Index: FFmpeg/libavfilter/vf_tonemap_videotoolbox.m
 +
 +static av_cold int tonemap_videotoolbox_pre_init(AVFilterContext *avctx)
```
`AVFilter.preinit =`

gnattu committed:
avfilter/tonemap_*: use preinit function

gnattu committed:
avfilter/tonemap_cuda: properly return value

gnattu commented:
```
2025-03-22T20:50:55.3826822Z  > [layer-45-harfbuzz 1/1] RUN --mount=src=scripts.d/45-harfbuzz.sh,dst=/stage.sh run_stage /stage.sh:
2025-03-22T20:50:55.3827576Z 5.945 compilation terminated.
2025-03-22T20:50:55.3828103Z 5.966 [13/81] Compiling C++ object src/libharfbuzz.a.p/hb-blob.cc.o
2025-03-22T20:50:55.3829244Z 5.966 FAILED: src/libharfbuzz.a.p/hb-blob.cc.o 
2025-03-22T20:50:55.3833313Z 5.966 x86_64-ffbuild-linux-gnu-g++ -Isrc/libharfbuzz.a.p -Isrc -I../src -I. -I.. -I/opt/ffbuild/include/freetype2 -I/opt/ffbuild/include -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++11 -fno-exceptions -O3 -fno-exceptions -fno-rtti -fno-threadsafe-statics -fvisibility-inlines-hidden -DHAVE_CONFIG_H -static-libgcc -static-libstdc++ -O2 -pipe -fPIC -DPIC -D_FORTIFY_SOURCE=2 -fstack-protector-strong -pthread -fvisibility=hidden -fno-semantic-interposition -fPIC -Wno-non-virtual-dtor -MD -MQ src/libharfbuzz.a.p/hb-blob.cc.o -MF src/libharfbuzz.a.p/hb-blob.cc.o.d -o src/libharfbuzz.a.p/hb-blob.cc.o -c ../src/hb-blob.cc
2025-03-22T20:50:55.3837288Z 5.966 In file included from ../src/hb-blob.cc:28:
2025-03-22T20:50:55.3837825Z 5.966 ../src/hb.hh:194:10: fatal error: cassert: No such file or directory
2025-03-22T20:50:55.3838218Z 5.966   194 | #include <cassert>
2025-03-22T20:50:55.3838926Z 5.966       |          ^~~~~~~~~
2025-03-22T20:50:55.3839268Z 5.966 compilation terminated.
2025-03-22T20:50:55.3839737Z 5.967 ninja: build stopped: subcommand failed.
```

Should we pin ct-ng to an earlier version for now?

nyanmisaka commented:
Weird, I didn't see this when I tested 7.1.1 last week. I'll rerun it and see.

nyanmisaka commented:
> Weird, I didn't see this when I tested 7.1.1 last week. I'll rerun it and see.

It does not fail in 7.1.1, where ct-ng as well as harfbuzz versions are updated.

nyanmisaka commented:
Tested and it fixed both gpu tonemap issues.
